### DISCUSSION TITLE
Having issue migrating from go-redis to valkey-go

### DISCUSSION THREAD

**Discussion by royatanu94** [valkey-io/valkey#797] [Q&A] [answered]
Currently we have used : 
rcOptions = &redis.UniversalOptions{
rc = redis.NewUniversalClient(rcOptions)  
and
func GetRedisClient(conf *RedisConfig) (redis.UniversalClient, error) 

What will be the compatible code for above two in valkey-go?

**Comment by zuiderkwast** [comment]
I don't know the answer to your question, but I know valkey-go is a fork of rueidis, not of go-redis. (Though there are discussions about forking go-redis too.)

@rueian FYI

**Reply by royatanu94** [reply]
@rueian Valkey still supports redis uri i.e. "redis://"  right?  E.g. redis://redis-cluster-password:7006": "redis-password"

**Reply by rueian** [reply]
Yes, it should support.

**Reply by royatanu94** [reply]
@rueian Is there any change in kpi metrics & logs format in valkey ? I assume log format kept same e.g. "1:M 23 Aug 2024 09:17:52.678 # Server initialized" but metrics are replaced with redis to valkey e.g. "redis_connected_slaves" to "valkey_connected_slaves" right? Is there any other change?

**Reply by rueian** [reply]
Hi @royatanu94, I believe you should ask the maintainers of https://github.com/oliver006/redis_exporter directly.

**Reply by royatanu94** [reply]
@rueian redis-exporter supports this since redis 7.0.15, but wanted to ensure any kpi metrics name explicitly changed in valkey or not, also  the log format of valkey server got changed or not.

**Reply by zuiderkwast** [reply]
@royatanu94 The log *format* didn't change. Most of the messages are the same too. We only changed some log messages which contain the name "redis".

What do you mean by KPI metrics in Valkey?

INFO, MEMORY STATS and similar commands didn't change. Only some field which are called "redis_version" or similar are changed in this way: We report the open source redis version we're compatible with and add another field for the valkey version.

**Comment by rueian** [comment]
Hi @royatanu94, @zuiderkwast 

The compatible code will be using valkeycompat:

```go
package main

import (
	"context"
	"time"
	"github.com/valkey-io/valkey-go"
	"github.com/valkey-io/valkey-go/valkeycompat"
)

func main() {
	ctx := context.Background()
	client, err := valkey.NewClient(valkey.ClientOption{InitAddress: []string{"127.0.0.1:6379"}})
	if err != nil {
		panic(err)
	}
	defer client.Close()

	compat := valkeycompat.NewAdapter(client)
	ok, _ := compat.SetNX(ctx, "key", "val", time.Second).Result()

	// with client side caching
	res, _ := compat.Cache(time.Second).Get(ctx, "key").Result()
}
```

The adapter is compatible with UniversalClient.

https://github.com/valkey-io/valkey-go/tree/main/valkeycompat

**Reply by royatanu94** [reply]
@rueian @zuiderkwast Thank you. I have already found this but in go-redis we have pubsub.go & universal.go from where we were using below methods
	Subscribe(ctx context.Context, channels ...string) *PubSub
        Publish(ctx context.Context, channel string, message interface{}) *IntCmd

I have found that publish is available in adapter.go but could not find Subscribe. Also could not find replacement of *redis.PubSub.
Please help to fix it.
Attaching the code written using go-redis which i am trying to migrate.



```
package utils

import (
	"context"
	"crypto/sha256"
	"encoding/base64"
	"math/rand"
	"time"

	"github.com/eapache/channels"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/redis/go-redis/v9"
	"google.golang.org/protobuf/proto"
)

const lockExpiration = time.Second * 5

var (
	PromMessageBusCounter = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Namespace: "lltest",
			Subsystem: "messagebus",
			Name:      "messages",
		},
		[]string{"type", "status"},
	)
)

func init() {
	prometheus.MustRegister(PromMessageBusCounter)
}

type MessageBus interface {
	Subscribe(ctx context.Context, channel string) (PubSub, error)
	// SubscribeQueue is like subscribe, but ensuring only a single instance gets to process the message
	SubscribeQueue(ctx context.Context, channel string) (PubSub, error)
	Publish(ctx context.Context, channel string, msg proto.Message) error
}

type PubSub interface {
	Channel() <-chan interface{}
	Payload(msg interface{}) []byte
	Close() error
}

type RedisMessageBus struct {
	rc redis.UniversalClient
}

func NewRedisMessageBus(rc redis.UniversalClient) MessageBus {
	return &RedisMessageBus{rc: rc}
}

func (r *RedisMessageBus) Lock(ctx context.Context, key string, expiration time.Duration) (bool, error) {
	return r.rc.SetNX(ctx, key, rand.Int(), expiration).Result()
}

func (r *RedisMessageBus) Subscribe(ctx context.Context, channel string) (PubSub, error) {
	ps := r.rc.Subscribe(ctx, channel)
	return &RedisPubSub{
		ps:   ps,
		c:    channels.Wrap(ps.Channel()).Out(),
		done: make(chan struct{}, 1),
	}, nil
}

func (r *RedisMessageBus) SubscribeQueue(ctx context.Context, channel string) (PubSub, error) {
	sub := r.rc.Subscribe(ctx, channel)
	c := make(chan *redis.Message, 100) // same chan size as redis pubsub
	ps := &RedisPubSub{
		ps:   sub,
		c:    channels.Wrap(c).Out(),
		done: make(chan struct{}, 1),
	}

	go func() {
		for {
			select {
			case <-ps.done:
				return
			case msg := <-sub.Channel():
				sha := sha256.Sum256([]byte(msg.Payload))
				hash := base64.StdEncoding.EncodeToString(sha[:])
				acquired, _ := r.Lock(ctx, hash, lockExpiration)
				if acquired {
					PromMessageBusCounter.WithLabelValues("in", "success").Add(1)
					c <- msg
				}
			}
		}
	}()

	return ps, nil
}

func (r *RedisMessageBus) Publish(ctx context.Context, channel string, msg proto.Message) error {
	b, err := proto.Marshal(msg)
	if err != nil {
		PromMessageBusCounter.WithLabelValues("out", "failure").Add(1)
		return err
	}

	err = r.rc.Publish(ctx, channel, b).Err()
	if err == nil {
		PromMessageBusCounter.WithLabelValues("out", "success").Add(1)
	} else {
		PromMessageBusCounter.WithLabelValues("out", "failure").Add(1)
	}

	return err
}

type RedisPubSub struct {
	ps   *redis.PubSub
	c    <-chan interface{}
	done chan struct{}
}

func (r *RedisPubSub) Channel() <-chan interface{} {
	return r.c
}

func (r *RedisPubSub) Payload(msg interface{}) []byte {
	return []byte(msg.(*redis.Message).Payload)
}

func (r *RedisPubSub) Close() error {
	r.done <- struct{}{}
	return r.ps.Close()
}

```

**Reply by rueian** [reply]
Thanks. I think the pubsub function will be landed in the adapter in the next release.

**Reply by royatanu94** [reply]
@rueian Can you please help how we can achieve above functionality using current available valkey-go code.

**Reply by rueian** [reply]
Hi @royatanu94,

You could try the pre-released [rueidiscompat/v1.0.42-alpha.2](https://github.com/redis/rueidis/releases/tag/rueidiscompat%2Fv1.0.42-alpha.2). There are `rueidiscompat.PubSub` and `rueidiscompat.Message` to serve as drop-in replacements. The implementation is here https://github.com/redis/rueidis/pull/592 for you to take a look at. The same functions will be landed in valkey-go once a stable version is released.

However, I would recommend you achieve the above functionality by just using the current `Receive` function of valkey-go: https://github.com/valkey-io/valkey-go?tab=readme-ov-file#pubsub

```go
err = client.Receive(context.Background(), client.B().Subscribe().Channel("ch1").Build(), func(msg valkey.PubSubMessage) {
    // handle the msg
})
```

**Reply by royatanu94** [reply]
@rueian Thank you. Can you please also help by providing alternatives in valkey-go for redis.Pipeliner, redis.Nil, redis.Script from go-redis.

**Reply by rueian** [reply]
Hi @royatanu94,

We have `valkey.Nil` and `valkey.NewLuaScript`. Do they meet your requirements?

A `Pipeliner` adapter will take me more time to implement due to its complexity. Before that, you may try the `DoMulti` instead. It is how we do pipelining manually: https://github.com/valkey-io/valkey-go?tab=readme-ov-file#manual-pipelining.

**Reply by royatanu94** [reply]
@rueian Thank you. I will try & let you know. Do we have any doc where I can find redis alternate in valkey? Now again I am looking for alternate of redis.Tx in valkey. Please let me know

**Reply by rueian** [reply]
We don’t have a full guide yet, but I am glad to help you here. Just ping me.

The Tx will come to the adapter after the Pipeliner since it is more complicated than the Pipeliner. As an alternative, you can make a transaction by doing this:

```go
client.Dedicated(func(c valkey.DedicatedClient) error {
    // watch keys first
    c.Do(ctx, c.B().Watch().Key("k1", "k2").Build())
    // perform read here
    c.Do(ctx, c.B().Mget().Key("k1", "k2").Build())
    // perform write with MULTI EXEC
    c.DoMulti(
        ctx,
        c.B().Multi().Build(),
        c.B().Set().Key("k1").Value("1").Build(),
        c.B().Set().Key("k2").Value("2").Build(),
        c.B().Exec().Build(),
    )
    return nil
})
```

Ref: https://github.com/valkey-io/valkey-go?tab=readme-ov-file#cas-transaction

**Reply by royatanu94** [reply]
@rueian Thank you for your help. Can you please share id from any instant messaging app where I can connect you, from your github I got mail id "rueiancsie@gmail.com"

**Reply by rueian** [reply]
Yes, but I would prefer we just use Github Issues or Discussions.

**Reply by royatanu94** [reply]
@rueian  what is alternate of UniversalClient.Watch in valkey-go adapter?

**Reply by rueian** [reply]
Hi @royatanu94, the current alternate of UniversalClient.Watch is exactly the same as above https://github.com/valkey-io/valkey-go?tab=readme-ov-file#cas-transaction.

**Reply by rueian** [reply, marked as answer]
Hi all,

The `Pipeliner`, `Tx`, `Nil`, `Script`, and `Watch` are all available in the [1.0.44](https://github.com/valkey-io/valkey-go/releases/tag/v1.0.44). Please refer to the https://github.com/valkey-io/valkey-go/blob/main/valkeycompat/README.md for more examples.

**Reply by royatanu94** [reply]
@rueian 
Hi,
In go-redis we used to get below logs when redis connect/disconnect happens.
```
redis: 2024/07/11 10:53:14 sentinel.go:706: sentinel: new master="master" addr="redis-1.redis.redis-pt-ns.svc.cluster.local:6379"
redis: 2024/07/11 10:53:14 pubsub.go:168: redis: discarding bad PubSub connection: read tcp 10.x.x.0:47226->10.244.x.x:6379: use of closed network connection
```

Do we have any logs for valkey-go?

**Reply by rueian** [reply]
Hi @royatanu94,

We don't have logging built-in. Defining a common logging interface in Go is a tough subject and we don't have a good practice yet. Some people don't even want logs at all https://github.com/redis/go-redis/issues/2555, https://github.com/redis/go-redis/issues/250. Currently, if you are interested in logging the `use of closed network connection`, you could do things like this:

```go
	client, err := valkey.NewClient(valkey.ClientOption{
		InitAddress: []string{"127.0.0.1:6379"},
		DialFn: func(addr string, dialer *net.Dialer, config *tls.Config) (conn net.Conn, err error) {
			conn, err = tls.DialWithDialer(dialer, "tcp", addr, config)
			conn = &LoggedConn{conn: conn} // log errors
			return conn, err
		},
	})
```
